src/codegen/query.cpp

```diff
@@ -31,40 +34,124 @@ void Query::Execute(std::unique_ptr<executor::ExecutorContext> executor_context,
 
   llvm::Type *runtime_state_type = runtime_state_.FinalizeType(codegen);
   size_t parameter_size = codegen.SizeOf(runtime_state_type);
-  PL_ASSERT((parameter_size % 8 == 0) && "parameter size not multiple of 8");
+  PL_ASSERT((parameter_size % 8 == 0) &&
+            parameter_size >= sizeof(FunctionArguments) &&
+            "parameter size not multiple of 8");
 
   // Allocate some space for the function arguments
   std::unique_ptr<char[]> param_data{new char[parameter_size]};
   char *param = param_data.get();
   PL_MEMSET(param, 0, parameter_size);
 
-  // We use this handy class to avoid complex casting and pointer manipulation
-  struct FunctionArguments {
-    storage::StorageManager *storage_manager;
-    executor::ExecutorContext *executor_context;
-    QueryParameters *query_parameters;
-    char *consumer_arg;
-    char rest[0];
-  } PACKED;
-
   // Set up the function arguments
   auto *func_args = reinterpret_cast<FunctionArguments *>(param_data.get());
   func_args->storage_manager = storage::StorageManager::GetInstance();
   func_args->executor_context = executor_context.get();
   func_args->query_parameters = &executor_context->GetParams();
   func_args->consumer_arg = consumer.GetConsumerState();
 
+  bool force_interpreter = settings::SettingsManager::GetBool(
+      settings::SettingId::codegen_interpreter);
+
+  if (is_compiled_ && !force_interpreter) {
+    ExecuteNative(func_args, stats);
+  } else {
+    try {
+      ExecuteInterpreter(func_args, stats);
+    } catch (interpreter::NotSupportedException e) {
+      LOG_ERROR("query not supported by interpreter: %s", e.what());
+
+      executor::ExecutionResult result;
+      result.m_result = ResultType::INVALID;
+      on_complete(result);
+      return;
+    }
+  }
+
+  executor::ExecutionResult result;
+  result.m_result = ResultType::SUCCESS;
+  result.m_processed = executor_context->num_processed;
+  on_complete(result);
+}
+
+void Query::Prepare(const LLVMFunctions &query_funcs) {
+  llvm_functions_ = query_funcs;
+
+  // verify the functions
+  // will also be done by Optimize() or Compile() if not done before,
+  // but we do not want to mix up the timings, so do it here
+  code_context_.Verify();
```

Comment:
You probably want to throw an exception here if verification fails.

Reply:
I changed `Query::Compile`, `CodeContext::Compile` and `CodeContext::Verify` to void and let `CodeContext::Verify` throw an Exception in case of errors. (We never expect it to fail) I also adapted `ASSERT_TRUE(code_context.Compile())` in some test cases.